dcos/recordio.py

```diff
@@ -0,0 +1,164 @@
+       arbitrary string of bytes into a 'RecordIO' message. Its
+       constructor takes a serialization function of the form
+       'serialize(message)'. This serialization function is
+       responsible for knowing how to take whatever message type
+       is passed to 'encode()' and serializing it to a string.
+
+       Once 'encode(message)' is called, it will use the
+       serialization function to convert 'message' into a string
+       of bytes, wrap that string fo bytes in a 'RecordIO' frame,
+       and return it.
+    """
+
+    def __init__(self, serialize):
+        """Initializes a 'RecordIO' encoder'.
+
+        ":param serialize: a function to serialize any message
+                           passed to 'encode()'
+        type serialize: function
+        """
+
+        self.serialize = serialize
+
+    def encode(self, message):
+        """Encode a message into 'RecordIO' format.
+
+        ":param message: a message to serialize and then wrap in
+                       a 'RecordIO' frame.
+        type message: object
+        returns: a serialzied message wrapped in a 'RecordIO' frame.
+        rtype: str
+        """
+
+        s = self.serialize(message)
+        return str(len(s)) + "\n" + s
+
+
+class Decoder(object):
+    """Decode a 'RecordIO' message back to an arbitrary message type.
+
+       This class encapsulates the process of decoding a message
+       previously encoded with 'RecordIO' back to an arbitrary
+       message type. Its
+       constructor takes a deserialization function of the form
+       'deserialize(bytes)'. This deserialization function is
+       responsible for knowing how to take a fully constructed
+       'RecordIO' message containing a string of bytes and
+       deserialize it back into the original message type.
+
+       The 'decode(data)' message takes a string of bytes as
+       input and buffers them across subsequent calls to
+       construct full 'RecordIO' messages that are then decoded
+       and returned in an array.
+    """
+    HEADER = 0
+    RECORD = 1
+    FAILED = 2
+
+    def __init__(self, deserialize):
+        """Initializes a 'RecordIO' decoder'.
+
+        ":param deserialize: a function to deserialize 'RecordIO'
+                             messages built up by subsequent
+                             calls to 'decode(data)'
+        type deserialize: function
+        """
+
+        self.deserialize = deserialize
+        self.state = self.HEADER
+        self.buffer = ""
+        self.length = 0
+
+    def decode(self, data):
+        """Decode data message 'RecordIO' format.
+
+        ":param data: a partial string of bytes making up a full
+                      'RecordIO' frame. Subsequent calls to this
+                      function maintain state to build up a full
+                      'RecordIO' message and decode it.
+        :type data : str
+        returns: a list of deserialized messages
+        rtype: list
+        """
```

Comment:
Again, are you saying that the constructor should not pass in a deserializer and that the return value should be a list of byte arrays that the caller then needs to do `map(deserialize, list)` over?